mllib/src/main/scala/org/apache/spark/ml/feature/TargetEncoder.scala

```diff
@@ -0,0 +1,460 @@
+import org.apache.spark.ml.attribute.NominalAttribute
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.param.shared._
+import org.apache.spark.ml.util._
+import org.apache.spark.sql.{Column, DataFrame, Dataset, Row}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+/** Private trait for params and common methods for TargetEncoder and TargetEncoderModel */
+private[ml] trait TargetEncoderBase extends Params with HasLabelCol
+  with HasInputCol with HasInputCols with HasOutputCol with HasOutputCols with HasHandleInvalid {
+
+  /**
+   * Param for how to handle invalid data during transform().
+   * Options are 'keep' (invalid data presented as an extra categorical feature) or
+   * 'error' (throw an error).
+   * Note that this Param is only used during transform; during fitting, invalid data
+   * will result in an error.
+   * Default: "error"
+   * @group param
+   */
+  @Since("4.0.0")
+  override val handleInvalid: Param[String] = new Param[String](this, "handleInvalid",
+    "How to handle invalid data during transform(). " +
+      "Options are 'keep' (invalid data presented as an extra categorical feature) " +
+      "or 'error' (throw an error). Note that this Param is only used during transform; " +
+      "during fitting, invalid data will result in an error.",
+    ParamValidators.inArray(TargetEncoder.supportedHandleInvalids))
+
+  setDefault(handleInvalid -> TargetEncoder.ERROR_INVALID)
+
+  @Since("4.0.0")
+  val targetType: Param[String] = new Param[String](this, "targetType",
+    "Type of label considered during fit(). " +
+      "Options are 'binary' and 'continuous'. When 'binary', estimates are calculated as " +
+      "conditional probability of the target given each category. When 'continuous', " +
+      "estimates are calculated as the average of the target given each category" +
+      "Note that this Param is only used during fitting.",
+    ParamValidators.inArray(TargetEncoder.supportedTargetTypes))
+
+  setDefault(targetType -> TargetEncoder.TARGET_BINARY)
+
+  final def getTargetType: String = $(targetType)
+
+  @Since("4.0.0")
+  val smoothing: DoubleParam = new DoubleParam(this, "smoothing",
+    "Smoothing factor for encodings. Smoothing blends in-class estimates with overall estimates " +
+      "according to the relative size of the particular class on the whole dataset, reducing the " +
+      "risk of overfitting due to unreliable estimates",
+    ParamValidators.gtEq(0.0))
+
+  setDefault(smoothing -> 0.0)
+
+  final def getSmoothing: Double = $(smoothing)
+
+  private[feature] lazy val inputFeatures = if (isSet(inputCol)) Array($(inputCol))
+  else if (isSet(inputCols)) $(inputCols)
+  else Array.empty[String]
+
+  private[feature] lazy val outputFeatures = if (isSet(outputCol)) Array($(outputCol))
+  else if (isSet(outputCols)) $(outputCols)
+  else inputFeatures.map{field: String => s"${field}_indexed"}
+
+  private[feature] def validateSchema(
```

Comment:
would you mind checking the scala style according to the section `Code style guide` in https://spark.apache.org/contributing.html?

Reply:
fixed